### DISCUSSION TITLE
support python 3.11

### DISCUSSION THREAD

**Discussion by LaurenzBeck** [ContinualAI/avalanche#1536] [Feature Request]
I would really appreciate the support of python 3.11
Currently, Version resolution fails with [poetry](https://python-poetry.org/), so I am not ably to install the library with my preferred stack.
Is there a technical reason why 3.11 is not supported yet, or was there simply no time/need yet?

thx for the good work and the awsome package!

**Comment by AntonioCarta** [comment]
Hi @LaurenzBeck, we are definitely interested in supporting 3.11. What is the error that you are getting? Are you able to install with pip instead?

I'm creating an issue to keep track of this.

**Reply by LaurenzBeck** [reply]
When trying to install avalanche-lib in a python 3.11 poetry environment I get:


```
  ❯ poetry add avalanche-lib
  Using version ^0.4.0 for avalanche-lib
  
  Updating dependencies
  Resolving dependencies... (0.2s)
  
  The current project's supported Python range (>=3.11,<3.12) is not compatible with some of the required packages Python requirement:
    - avalanche-lib requires Python >=3.7,<3.11, so it will not be satisfied for Python >=3.11,<3.12
  
  Because no versions of avalanche-lib match >0.4.0,<0.5.0
   and avalanche-lib (0.4.0) requires Python >=3.7,<3.11, avalanche-lib is forbidden.
  So, because <project name> depends on avalanche-lib (^0.4.0), version solving failed.
  
    • Check your dependencies Python requirement: The Python requirement can be specified via the python or markers properties
  
      For avalanche-lib, a possible solution would be to set the python property to "<empty>"
  
      https://python-poetry.org/docs/dependency-specification/#python-restricted-dependencies,
      https://python-poetry.org/docs/dependency-specification/#using-environment-markers

```

Trying to circumvent the stricter dependecy resolution of poetry by installing it manually with pip leads to:


```
  ❯ pip install avalanche-lib
  Looking in indexes: <company internal pypi mirror>
  ERROR: Ignored the following versions that require a different python version: 0.1.0 Requires-Python >=3.6,<3.10; 0.2.0 Requires-Python >=3.6,<3.11; 0.2.1 Requires-Python >=3.6,<3.11; 0.3.0 Requires-Python >=3.7,<3.11; 0.3.1 Requires-Python >=3.7,<3.11; 0.4.0 Requires-Python >=3.7,<3.11
  ERROR: Could not find a version that satisfies the requirement avalanche-lib (from versions: none)
  ERROR: No matching distribution found for avalanche-lib

```

I suspected that the codebase should work for Python 3.11 without problems, I tried forking the repository and relaxing/adapting the `` version constraint in https://github.com/ContinualAI/avalanche/blob/master/setup.py#L63 , but installing the fork with did also not work:


```
  ❯ poetry add git+https://github.com/LaurenzBeck/avalanche.git
  
  Updating dependencies
  Resolving dependencies... (0.0s)
  
  The current project's supported Python range (>=3.11,<3.12) is not compatible with some of the required packages Python requirement:
    - avalanche-lib requires Python >=3.7,<3.11, so it will not be satisfied for Python >=3.11,<3.12
  
  Because <project name> depends on avalanche-lib (0.4.0a0) @ git+https://github.com/LaurenzBeck/avalanche.git@HEAD which requires Python >=3.7,<3.11, version solving failed.

```

I am not an expert on this topic though, so excuse any silly mistakes :)

**Comment by Daniel451** [comment]
@AntonioCarta @LaurenzBeck is there any update on this issue? Since PyTorch 2.1 is now fully supporting 3.11 as well as torchvision, ... and 3.11 has performance gains over 3.10 it would be very nice to have 3.11 available for Avalanche, too.

**Reply by LaurenzBeck** [reply]
Not from my side, I switched to a 3.10 environment for my evaluation of avalanche.
As far as I can tell, supporting 3.11. is probably just a matter of testing and deployment, as there were very few actionable changes.
From my experience, such "simple" tasks can still take a few days. So Antonio is probably waiting for the next release where he has to do the testing and deployment process anyways.

**Comment by AntonioCarta** [comment]
@LaurenzBeck @Daniel451 The master branch now supports 3.11. Avalanche 0.5.0 will be realeased shortly (possibly today).
